### PR TITLE
Npm: Add support for declared authors

### PR DESCRIPTION
**Pull request by oheger-bosch**
Extract information about the author of a package from package.json
and make it available in the metadata of packages and projects.

Note: The major part of the changes is related to adapting the expected output of tests.

### CODE REVIEW

File: analyzer/src/funTest/assets/projects/synthetic/npm/package-lock/package.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
     "url": "https://github.com/oss-review-toolkit/ort.git"
   },
   "license": "Apache-2.0",
+  "author": "The Author <the.author@copy-right.com> (https://www.the.author.org/index.html)",
```
Comment (tsteenbe) [inline analyzer/src/funTest/assets/projects/synthetic/npm/package-lock/package.json:10]: If you using fake domain please use example.com or oss-review-toolkit.org. Who know what someone may host on copy-right.com or author.org

**Review by tsteenbe — DISMISSED** [review]

File: analyzer/src/funTest/assets/projects/synthetic/npm/package-lock/package.json (hunk 1 above)
Comment (maxhbr) [reply analyzer/src/funTest/assets/projects/synthetic/npm/package-lock/package.json:10]: there are explicit tlds for this use case, e.g. `.test`, `.example` and `.invalid`. See https://tools.ietf.org/html/rfc2606

File: analyzer/src/main/kotlin/managers/Npm.kt [hunk 2]
```diff
@@ -183,6 +185,31 @@ open class Npm(
         }
     }
 
+    /**
+     * Parse information about the author. According to
+     * https://docs.npmjs.com/files/package.json#people-fields-author-contributors, there are two formats to
+     * specify the author of a package: An object with multiple properties or a single string.
+     */
+    private fun parseAuthor(json: JsonNode): SortedSet<String> =
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Npm.kt:193]: Shouldn't this better simply return a (nullable) `String` as the set only ever contains at most a single element?

File: analyzer/src/funTest/assets/projects/synthetic/npm/package-lock/package.json (hunk 1 above)
Comment (oheger-bosch) [reply analyzer/src/funTest/assets/projects/synthetic/npm/package-lock/package.json]: Changed mail addresses and URLs in two package.json files.

File: analyzer/src/main/kotlin/managers/Npm.kt [hunk 3]
```diff
@@ -183,6 +185,31 @@ open class Npm(
...
+                }
+            }?.let { add(it) }
+        }
+
+    /**
+     * Parse the author if it is defined as a single string in the format "Name <mail> (url)". Try to extract the name
+     * part from the string or return the full string if no mail or url components are found.
+     */
+    private fun parseAuthorString(authorStr: String?): String? = authorStr?.let { str ->
+        val posSeparator = min(str.findSeparator('<'), str.findSeparator('('))
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Npm.kt:209]: How about instead leveraging `substringBefore()`, or maybe better yet, `split('<', '(', limit = 2)`, to be able to get rid of the `findSeparator` helpfer function?

File: analyzer/src/funTest/assets/projects/synthetic/npm/README.md [hunk 4]
```diff
@@ -1,6 +1,6 @@
 # npm
 
-This repository contains multiple NPM projects for testing. Each directory contains the same package.json plus the
+This repository contains multiple NPM projects for testing. Each directory contains an equivalent package.json plus the
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/synthetic/npm/README.md:3]: Nit: This should go to a separate commit.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/Npm.kt (hunk 2 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/Npm.kt:193]: For the calling code it is more convenient that this function returns a set because this is needed for the package metadata. However, it might then be more suitable to rename the function to `parseAuthors()`. WDYT?

File: analyzer/src/funTest/assets/projects/synthetic/npm/README.md (hunk 4 above)
Comment (oheger-bosch) [reply analyzer/src/funTest/assets/projects/synthetic/npm/README.md:3]: I think this change belongs to the same commit: Originally, there were multiple identical `package.json` files. This commit changed some of them to contain author metadata in different formats. Hence, the README is no longer correct - the `package.json` files are not identical, but produce the same outcome -, and it has been adjusted.

Comment (sschuberth) [reply analyzer/src/funTest/assets/projects/synthetic/npm/README.md:3]: You're right. I was looking at a wrong diff which made it seem as if the author field in general was there before.

File: analyzer/src/main/kotlin/managers/Npm.kt (hunk 3 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/Npm.kt]: Implemented using `substringBefore()`.

File: analyzer/src/main/kotlin/managers/Npm.kt (hunk 2 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Npm.kt:191]: When keeping the implementation as-is then yes, we should rename the function to plural `parseAuthors()`. But I'm still wondering whether instead writing `sortedSetOf(parseAuthor())` at each call site is really that bad.

Maybe a (quite academic) argument for using plural `parseAuthors()` would be that we then could easily extend the function if NPM would start to support author arrays, similar to like we already do the parsing in `parseLicenses()`.

I'll leave the decision up to you.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/Npm.kt (hunk 2 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/managers/Npm.kt:191]: I went for `parseAuthors()`. Writing `sortedSetOf(parseAuthor())` on the caller side would not be sufficient because the case of a *null* result still has to be handled.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Npm.kt]: Right. While there is `setOfNotNull()`, there's unfortunately not `sortedSetOfNotNull()`.

**Review by sschuberth — APPROVED** [review]
